src/lib/png/png.c

```diff
@@ -0,0 +1,976 @@
+
+	int bytewidth = (bpp + 7) / 8;
+	int linebytes = (w * bpp + 7) / 8;
+
+	for (i = 0; i < h; i++) {
+		int outindex = linebytes * i;
+		int inindex = (1 + linebytes) * i;
+		uint8_t filterType = in[inindex];
+
+		if (unfilter_second_step(&out[outindex], &in[inindex + 1],
+			prevline, bytewidth, filterType, linebytes) == -1) {
+				printf("Error in filter\n");
+				return -1;
+			}
+
+		prevline = &out[outindex];
+	}
+	return 0;
+}
+
+/**
+ * 1)This function converts the filtered-padded-interlaced data in img
+ * 2)out: big enough for contain full image,
+ * 3)in: contain the full decompressed data from
+ * the IDAT chunks (with filter index bytes and possible padding bits)
+ * 4)Steps:
+ * a) if no Adam7: 1) unfilter 2) TODO: remove padding bits (if bpp < 8)
+ * b) TODO: if adam7: 1) 7x unfilter 2) 7x remove padding bits 3) Adam7_deinterlace
+ * return 0 - success or error
+*/
+static int after_decode_unfilter(uint8_t *out, uint8_t *in, int w, int h, int interflace, int bpp) {
+	if (interflace == 0) {
+		/* TODO removePaddingBits */
+		unfilter_first_step(out, in, w, h, bpp);
+	} else {
+		printf("TODO : Adam7 not supported yet\n");
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * this function start of decoding IDAT
+ * out : buffer for decoded image, big enough for img
+ * in : contain IDAT bytes
+ * inlen : length of in(without zlib_head)
+*/
+static int inflate(uint8_t *out, uint8_t *in, int inlen, struct png_header *png_head) {
+	int bp = 3;/* bit point in in */
+	int pos = 0;/*pos in out */
+	int btype = 2;/* huffman type (1 - fixed/2 - dynamic) */
+	int bpp;
+	uint8_t *Huffman_out = (uint8_t*)malloc(png_head->width * png_head->height * 5);
+
+	if (inflate_huffman(Huffman_out, in, &bp, &pos, inlen, btype) != 0) {
+		printf("Error : can't decode IDAT chunk.\n");
+		return -1;
+	}
+
+	bpp = png_head->color_type * png_head->bit_depth;
+
+	if (after_decode_unfilter(out, Huffman_out, png_head->width, png_head->height,
+												png_head->interlace_method, bpp) != 0) {
+		printf("Error while unfilter\n");
+		return -1;
+	}
+	return 0;
+}
+
+static int read_IDAT(int fd, uint8_t *image, int len_IDAT, struct png_header *png_head) {
+	uint8_t zlib[2];
+
+	len_IDAT -= 2; /* IDAT len minus len of zlib header */
+
+	uint8_t *data_IDAT = (uint8_t*)malloc(len_IDAT);
+
+	/* read zlib header */
+	if (read(fd, zlib, sizeof(zlib)) != sizeof(zlib)) {
+		printf("Error : can't read zlib info\n");
+		return -1;
+	}
+
+	/* read IDAT chunk */
+	if (read(fd, data_IDAT, len_IDAT) != len_IDAT) {
+		printf("Error : can't read IDAT chunk\n");
+		return -1;
+	}
+
+	if (inflate(image, data_IDAT, len_IDAT, png_head) != 0) {
+		return -1;
+	}
+	return 0;
+}
+
+/* Read IHDR chunk and get info */
+static int read_IHDR(int fd, struct png_header *png_head) {
+
+	uint32_t img_width, img_height;
+	uint32_t crc; 	       		/* CRC-32 */
+
+	/* read width of image ant put in header struct */
+	if (read(fd, &img_width, sizeof(img_width)) != sizeof(img_width)) {
```

Comment:
Why don't read the whole header with `read(fd, &png_head, sizeof(png_head))`?

This way this function will be much shorter.

Reply:
some variables in header have already converted from bytes value and just read the header and then each time convert will be inconvenient

Reply:
reworked this fun